source/common/router/router.cc

```diff
@@ -837,6 +842,35 @@ void Filter::sendNoHealthyUpstreamResponse() {
                              StreamInfo::ResponseCodeDetails::get().NoHealthyUpstream);
 }
 
+void Filter::giveUpRetryAndShadow(bool drain_buffer) {
+  cluster_->trafficStats()->retry_or_shadow_abandoned_.inc();
+  retry_state_.reset();
+  active_shadow_policies_.clear();
+  if (buffer_limit_exceeded_no_connection_) {
+    callbacks_->onDecoderFilterBelowWriteBufferLowWatermark();
+  }
+  request_buffer_overflowed_ = true;
+  buffer_limit_exceeded_no_connection_ = false;
+  if (drain_buffer) {
+    // we previously went over the buffer limit, which pause the read on the downstream connection.
+    // draining the buffer so we enable read on the downstream connection;
+    // the buffer size will go bellow
+    // this buffer is only used to send the body when retrying and shadowing which has been disable
+    // above.
+    callbacks_->modifyDecodingBuffer([](Buffer::Instance& data) { data.drain(data.length()); });
+  }
+}
+
+void Filter::onUpstreamConnectionEstablished() {
+  wait_for_connect_ = false;
+  ENVOY_LOG(trace, "Got connection with upstream");
+  if (buffer_limit_exceeded_no_connection_) {
+    ENVOY_LOG(debug, "Giving up on retry and shadow and enabling read on downstream connection");
+    // disabling retry and shadow: the requests is to big
+    giveUpRetryAndShadow(true);
```

Comment:
IIUC, are we just deferring giving up retry and shadow until either (1) out of retry attempts or (2) we have a connection established?

Prior we'd cancel retries if the shadow limit was reached immediately and send a local reply